runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -8004,35 +8007,87 @@ class INTERPRETER_CLASS
 				goto retry;
 			}
 		}
+
 		return rc;
-#else
-	Assert_VM_unreachable();
-	return EXECUTE_BYTECODE;
-#endif /* J9VM_OPT_METHOD_HANDLE */
 	}
+#elif defined(J9VM_OPT_OPENJDK_METHODHANDLE) /* defined(J9VM_OPT_METHOD_HANDLE) */
+	VMINLINE VM_BytecodeAction
+	invokedynamic(REGISTER_ARGS_LIST)
+	{
+retry:
+		VM_BytecodeAction rc = GOTO_RUN_METHOD;
+		U_16 index = *(U_16 *)(_pc + 1);
+
+		J9ConstantPool *ramConstantPool = J9_CP_FROM_METHOD(_literals);
+		J9InvokeCacheEntry *invokeCache = ((J9InvokeCacheEntry *)ramConstantPool->ramClass->callSites) + index;
```

Comment:
I don't see the definition of `J9InvokeCacheEntry` in this PR or in the already committed code.  Am I missing it?

Hard to tell if the barriers are correct in this path without seeing that code and the updates to resolveInvokeDynmaic

Reply:
```
typedef struct J9InvokeCacheEntry {
	j9object_t target;
	j9object_t appendix;
} J9InvokeCacheEntry;
```

`J9InvokeCacheEntry` is planned to be added in the `resolvesupport.cpp` changes (separate PR).

Reply:
We read both values from the cache here without a memory barrier.  How do we ensure that both are visible at this point?  Are they always written such that:
```
cache->appendix = ....;
<membarrier>
cache->target = ....;
```
so that the store of appendix can't be moved passed the store of target or another racing thread can read incompletely resolved values

Reply:
Can you link the separate PR here?  We need to review both parts together to tell if there's a problem.

Also, multiple threads can race to resolve an invokedynamic.  The old resolve code used a CAS to resolve the race.  How is that handled with the two slot solution?

Reply:
In `resolvesupport.cpp`, @fengxue-IS has added sync code. The sync code will assure that both values are written before any thread reads them. 

`target` will be `NULL` until the resolution is complete. So, all threads will take the resolve path if `target` is `NULL`. In the resolve path, there will be a spin loop which will prevent the threads to proceed until the resolved values are successfully written. `target` will be written last to indicate the completion of resolution.

@fengxue-IS thumbs up to confirm?

Reply:
> Can you link the separate PR here?

yes, the resolution PR will be linked with this PR. it still needs to be created.

> The old resolve code used a CAS to resolve the race. How is that handled with the two slot solution?

Similar spin loop (CAS) approach, the last slot to be written is used as an indicator for completion of resolution.